core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/clustering/hazelcast/kubernetes/KubernetesMembershipScheme.java

```diff
@@ -0,0 +1,164 @@
+
+    }
+
+    @Override
+    public void setCarbonCluster(HazelcastCarbonClusterImpl hazelcastCarbonCluster) {
+
+        this.carbonCluster = hazelcastCarbonCluster;
+    }
+
+    @Override
+    public void init() {
+
+        String localMemberPort = "4000";
+        Parameter namespace = getParameter(KubernetesConstants.NAMESPACE_PROPERTY);
+        Parameter serviceName = getParameter(KubernetesConstants.SERVICE_NAME_PROPERTY);
+
+        log.info("Initializing kubernetes membership scheme...");
+        nwConfig.getJoin().getMulticastConfig().setEnabled(false);
+        nwConfig.getJoin().getAwsConfig().setEnabled(false);
+        nwConfig.getJoin().getKubernetesConfig().setEnabled(true);
+
+        Parameter localMemberPortParam = getParameter(HazelcastConstants.LOCAL_MEMBER_PORT);
+        if (localMemberPortParam != null) {
+            localMemberPort = ((String) localMemberPortParam.getValue()).trim();
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_PORT, localMemberPort);
+        } else {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_PORT, localMemberPort);
+        }
+
+        if (namespace != null) {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.NAMESPACE, ((String) namespace.getValue()).trim());
+        }
+        if (serviceName != null) {
+            nwConfig.getJoin().getKubernetesConfig().setProperty(KubernetesConstants.SERVICE_NAME, ((String) serviceName.getValue()).trim());
+        }
+    }
+
+    @Override
+    public void joinGroup() {
+
+        primaryHazelcastInstance.getCluster().addMembershipListener(new KubernetesMembershipSchemeListener());
+    }
+
+    private Parameter getParameter(String name) {
+
+        return parameters.get(name);
+    }
+
+    /**
+     * Kubernetes membership scheme listener
+     */
+    private class KubernetesMembershipSchemeListener implements MembershipListener {
+
+        @Override
+        public void memberAdded(MembershipEvent membershipEvent) {
+
+            Member member = membershipEvent.getMember();
+            TcpIpConfig tcpIpConfig = nwConfig.getJoin().getTcpIpConfig();
+            List<String> memberList = tcpIpConfig.getMembers();
+            if (!memberList.contains(member.getSocketAddress().getAddress().getHostAddress())) {
+                tcpIpConfig.addMember(String.valueOf(member.getSocketAddress().getAddress().getHostAddress()));
+            }
+
+            // Send all cluster messages.
+            carbonCluster.memberAdded(member);
+            log.info(String.format("Member joined: [UUID] %s, [Address] %s", member.getUuid(),
+                    member.getSocketAddress().toString()));
+            // Wait for sometime for the member to completely join before replaying messages.
+            try {
+                Thread.sleep(5000);
+            } catch (InterruptedException ignored) {
```

Comment:
What will happen when a thread interrupt occurs?

Reply:
Here I have moved the kubernetes membership scheme code base into the kernel with some improvements. https://github.com/wso2/kubernetes-common/tree/master/kubernetes-membership-scheme

As per the code, it ignores the thread interrupt and send message to the newly added member node.